{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "811cb3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "from scipy import stats\n",
    "import scvelo as scv\n",
    "import pandas as pd\n",
    "from sim_low_umi import downsample_anndata\n",
    "import gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f37a90df",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = Path(\"/root/autodl-tmp/dataset\")\n",
    "DATASET = \"gastrulation_downsample\"\n",
    "(DATA_DIR / DATASET / \"processed\").mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "158097b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9815 × 53801\n",
       "    obs: 'sample', 'stage', 'sequencing.batch', 'theiler', 'celltype', 'clusters'\n",
       "    var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'MURK_gene', 'Δm', 'scaled Δm'\n",
       "    uns: 'celltype_colors'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    layers: 'spliced', 'unspliced'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(DATA_DIR / DATASET / \"raw\" / \"MouseErythroid_anndata.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5135199e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered out 47456 genes that are detected 20 counts (shared).\n",
      "Normalized count data: X, spliced, unspliced.\n",
      "Extracted 2000 highly variable genes.\n",
      "Logarithmized X.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.\n",
      "  log1p(adata)\n"
     ]
    }
   ],
   "source": [
    "adata.layers['raw_spliced'] = adata.layers['spliced'].copy()\n",
    "adata.layers['raw_unspliced'] = adata.layers['unspliced'].copy()\n",
    "scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0284656",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered out 1046 genes that are detected 20 counts (shared).\n",
      "WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.\n",
      "Normalized count data: spliced, unspliced.\n",
      "Skip filtering by dispersion since number of variables are less than `n_top_genes`.\n",
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.\n",
      "  log1p(adata)\n",
      "C:\\Users\\18307\\AppData\\Local\\Temp\\ipykernel_46304\\37169063.py:17: DeprecationWarning: Automatic neighbor calculation is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors first with Scanpy.\n",
      "  scv.pp.moments(downsampled_adata, n_neighbors=30, n_pcs=30)\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\moments.py:71: DeprecationWarning: `neighbors` is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors with Scanpy.\n",
      "  neighbors(\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\neighbors.py:233: DeprecationWarning: Automatic computation of PCA is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute PCA with Scanpy first.\n",
      "  _set_pca(adata=adata, n_pcs=n_pcs, use_highly_variable=use_highly_variable)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logarithmized X.\n",
      "computing neighbors\n",
      "    finished (0:00:53) --> added \n",
      "    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)\n",
      "computing moments based on connectivities\n",
      "    finished (0:00:00) --> added \n",
      "    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)\n",
      "Filtered out 613 genes that are detected 20 counts (shared).\n",
      "WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.\n",
      "Normalized count data: spliced, unspliced.\n",
      "Skip filtering by dispersion since number of variables are less than `n_top_genes`.\n",
      "WARNING: adata.X seems to be already log-transformed.\n",
      "Logarithmized X.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.\n",
      "  log1p(adata)\n",
      "C:\\Users\\18307\\AppData\\Local\\Temp\\ipykernel_46304\\37169063.py:17: DeprecationWarning: Automatic neighbor calculation is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors first with Scanpy.\n",
      "  scv.pp.moments(downsampled_adata, n_neighbors=30, n_pcs=30)\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\moments.py:71: DeprecationWarning: `neighbors` is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors with Scanpy.\n",
      "  neighbors(\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\neighbors.py:233: DeprecationWarning: Automatic computation of PCA is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute PCA with Scanpy first.\n",
      "  _set_pca(adata=adata, n_pcs=n_pcs, use_highly_variable=use_highly_variable)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing neighbors\n",
      "    finished (0:00:05) --> added \n",
      "    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)\n",
      "computing moments based on connectivities\n",
      "    finished (0:00:01) --> added \n",
      "    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)\n",
      "Filtered out 362 genes that are detected 20 counts (shared).\n",
      "WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.\n",
      "Normalized count data: spliced, unspliced.\n",
      "Skip filtering by dispersion since number of variables are less than `n_top_genes`.\n",
      "WARNING: adata.X seems to be already log-transformed.\n",
      "Logarithmized X.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.\n",
      "  log1p(adata)\n",
      "C:\\Users\\18307\\AppData\\Local\\Temp\\ipykernel_46304\\37169063.py:17: DeprecationWarning: Automatic neighbor calculation is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors first with Scanpy.\n",
      "  scv.pp.moments(downsampled_adata, n_neighbors=30, n_pcs=30)\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\moments.py:71: DeprecationWarning: `neighbors` is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors with Scanpy.\n",
      "  neighbors(\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\neighbors.py:233: DeprecationWarning: Automatic computation of PCA is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute PCA with Scanpy first.\n",
      "  _set_pca(adata=adata, n_pcs=n_pcs, use_highly_variable=use_highly_variable)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing neighbors\n",
      "    finished (0:00:04) --> added \n",
      "    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)\n",
      "computing moments based on connectivities\n",
      "    finished (0:00:01) --> added \n",
      "    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)\n",
      "Filtered out 154 genes that are detected 20 counts (shared).\n",
      "WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.\n",
      "Normalized count data: spliced, unspliced.\n",
      "Skip filtering by dispersion since number of variables are less than `n_top_genes`.\n",
      "WARNING: adata.X seems to be already log-transformed.\n",
      "Logarithmized X.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.\n",
      "  log1p(adata)\n",
      "C:\\Users\\18307\\AppData\\Local\\Temp\\ipykernel_46304\\37169063.py:17: DeprecationWarning: Automatic neighbor calculation is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors first with Scanpy.\n",
      "  scv.pp.moments(downsampled_adata, n_neighbors=30, n_pcs=30)\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\moments.py:71: DeprecationWarning: `neighbors` is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors with Scanpy.\n",
      "  neighbors(\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\neighbors.py:233: DeprecationWarning: Automatic computation of PCA is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute PCA with Scanpy first.\n",
      "  _set_pca(adata=adata, n_pcs=n_pcs, use_highly_variable=use_highly_variable)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing neighbors\n",
      "    finished (0:00:04) --> added \n",
      "    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)\n",
      "computing moments based on connectivities\n",
      "    finished (0:00:01) --> added \n",
      "    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)\n",
      "WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.\n",
      "Normalized count data: spliced, unspliced.\n",
      "Extracted 2000 highly variable genes.\n",
      "WARNING: adata.X seems to be already log-transformed.\n",
      "Logarithmized X.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.\n",
      "  log1p(adata)\n",
      "C:\\Users\\18307\\AppData\\Local\\Temp\\ipykernel_46304\\37169063.py:17: DeprecationWarning: Automatic neighbor calculation is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors first with Scanpy.\n",
      "  scv.pp.moments(downsampled_adata, n_neighbors=30, n_pcs=30)\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\moments.py:71: DeprecationWarning: `neighbors` is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute neighbors with Scanpy.\n",
      "  neighbors(\n",
      "d:\\softwares\\anaconda3\\envs\\scvelo_py39\\lib\\site-packages\\scvelo\\preprocessing\\neighbors.py:233: DeprecationWarning: Automatic computation of PCA is deprecated since scvelo==0.4.0 and will be removed in a future version of scVelo. Please compute PCA with Scanpy first.\n",
      "  _set_pca(adata=adata, n_pcs=n_pcs, use_highly_variable=use_highly_variable)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing neighbors\n",
      "    finished (0:00:04) --> added \n",
      "    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)\n",
      "computing moments based on connectivities\n",
      "    finished (0:00:01) --> added \n",
      "    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)\n"
     ]
    }
   ],
   "source": [
    "# 1.0 is the original data, 0.2 is 20% of the original data, etc.\n",
    "for idx, downsample_rate in enumerate([0.2, 0.4, 0.6, 0.8, 1.0]):\n",
    "    downsampled_adata = downsample_anndata(\n",
    "        adata,\n",
    "        downsample_rate=downsample_rate,\n",
    "        seed=0\n",
    "    )\n",
    "    downsampled_adata.layers['raw_spliced'] = downsampled_adata.layers['spliced']\n",
    "    downsampled_adata.layers['raw_unspliced'] = downsampled_adata.layers['unspliced']\n",
    "    scv.pp.filter_and_normalize(downsampled_adata, min_shared_counts=20, n_top_genes=2000)\n",
    "    if 'X_pca' in downsampled_adata.obsm:\n",
    "        del downsampled_adata.obsm['X_pca']\n",
    "    if 'pca' in downsampled_adata.uns:\n",
    "        del downsampled_adata.uns['pca']\n",
    "    if \"neighbors\" in downsampled_adata.uns.keys():\n",
    "        del downsampled_adata.uns['neighbors']\n",
    "    scv.pp.moments(downsampled_adata, n_neighbors=30, n_pcs=30)\n",
    "    sc.tl.umap(downsampled_adata, random_state=0)\n",
    "    downsampled_adata.obs['u_lib_size_raw'] = downsampled_adata.layers['raw_unspliced'].toarray().sum(-1) \n",
    "    downsampled_adata.obs['s_lib_size_raw'] = downsampled_adata.layers['raw_spliced'].toarray().sum(-1)\n",
    "    downsampled_adata.write_h5ad(DATA_DIR / DATASET / \"processed\" / f\"adata_preprocessed_{idx}.h5ad\")\n",
    "    del downsampled_adata\n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7c55d4c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downsample rate: 0.2, Mean spliced expression: 0.47\n",
      "Downsample rate: 0.4, Mean spliced expression: 0.68\n",
      "Downsample rate: 0.6, Mean spliced expression: 0.88\n",
      "Downsample rate: 0.8, Mean spliced expression: 1.05\n",
      "Downsample rate: 1.0, Mean spliced expression: 1.23\n"
     ]
    }
   ],
   "source": [
    "for idx, downsample_rate in enumerate([0.2, 0.4, 0.6, 0.8, 1.0]):\n",
    "    adata = sc.read_h5ad(DATA_DIR / DATASET / \"processed\" / f\"adata_preprocessed_{idx}.h5ad\")\n",
    "    spliced = adata.layers['raw_spliced'].toarray().astype(np.float32)\n",
    "    mean_expression = np.mean(spliced)\n",
    "    print(f\"Downsample rate: {downsample_rate}, Mean spliced expression: {mean_expression:.2f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "evaluation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
